// ionic/dist/commands/ssh/add.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const path = require("path");
const chalk = require("chalk");
const cli_utils_1 = require("@ionic/cli-utils");
const base_1 = require("./base");
let SSHAddCommand = class SSHAddCommand extends base_1.SSHBaseCommand {
    run(inputs, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { ERROR_SSH_ANNOTATION_INVALID_WHITESPACE, ERROR_SSH_ANNOTATION_MISSING, ERROR_SSH_INVALID_PUBKEY, parsePublicKeyFile, } = yield Promise.resolve().then(function () { return require('../../lib/ssh'); });
            const pubkeyPath = path.resolve(inputs[0]);
            const pubkeyName = cli_utils_1.prettyPath(pubkeyPath);
            let pubkey;
            try {
                [pubkey, , ,] = yield parsePublicKeyFile(pubkeyPath);
            }
            catch (e) {
                if (e === cli_utils_1.ERROR_FILE_NOT_FOUND) {
                    throw new cli_utils_1.FatalException(`${chalk.bold(cli_utils_1.prettyPath(pubkeyPath))} does not appear to exist. Please specify a valid SSH public key.\n` +
                        `If you are having issues, try using ${chalk.green('ionic ssh setup')}.`);
                }
                else if (e === ERROR_SSH_INVALID_PUBKEY) {
                    throw new cli_utils_1.FatalException(`${chalk.bold(pubkeyName)} does not appear to be a valid SSH public key. (Not in ${chalk.bold('authorized_keys')} file format.)\n` +
                        `If you are having issues, try using ${chalk.green('ionic ssh setup')}.`);
                }
                else if (e === ERROR_SSH_ANNOTATION_MISSING) {
                    throw new cli_utils_1.FatalException(`${chalk.bold(pubkeyName)} is missing an annotation/comment after the public key.\n` +
                        `If you are using ${chalk.green('ssh-keygen')}, try using the ${chalk.green('-C')} flag.\n` +
                        `If you are having issues, try using ${chalk.green('ionic ssh setup')}.`);
                }
                else if (e === ERROR_SSH_ANNOTATION_INVALID_WHITESPACE) {
                    throw new cli_utils_1.FatalException(`${chalk.bold(pubkeyName)} has an annotation/comment that has whitespace.\n` +
                        `Try changing the comment to something more like an identifier.\n` +
                        `If you are having issues, try using ${chalk.green('ionic ssh setup')}.`);
                }
                throw e;
            }
            const config = yield this.env.config.load();
            const token = yield this.env.session.getUserToken();
            const req = this.env.client.make('POST', `/users/${config.user.id}/sshkeys`)
                .set('Authorization', `Bearer ${token}`)
                .send({ pubkey });
            try {
                const res = yield this.env.client.do(req);
                if (!cli_utils_1.isSSHKeyResponse(res)) {
                    throw cli_utils_1.createFatalAPIFormat(req, res);
                }
                const words = res.meta.status === 201 ? 'added to' : 'updated on';
                this.env.log.ok(`Your public key (${chalk.bold(res.data.id)}) has been ${words} Ionic!`);
            }
            catch (e) {
                if (cli_utils_1.isSuperAgentError(e) && e.response.status === 409) {
                    this.env.log.info('Pubkey already added to Ionic.');
                }
                else {
                    throw e;
                }
            }
        });
    }
};
SSHAddCommand = tslib_1.__decorate([
    cli_utils_1.CommandMetadata({
        name: 'add',
        type: 'global',
        backends: [cli_utils_1.BACKEND_PRO],
        description: 'Add an SSH public key to Ionic',
        inputs: [
            {
                name: 'pubkey-path',
                description: 'Location of public key file to add to Ionic',
                validators: [cli_utils_1.validators.required],
            }
        ],
    })
], SSHAddCommand);
exports.SSHAddCommand = SSHAddCommand;
